lang/src/main/kotlin/org/partiql/lang/eval/ExprValue.kt

```diff
@@ -67,21 +74,284 @@ interface ExprValue : Iterable<ExprValue>, Faceted {
     override operator fun iterator(): Iterator<ExprValue>
 
     companion object {
+        // Constructor functions
+        private fun constructMissingValue() =
+            object : BaseExprValue() {
+                override val type = ExprValueType.MISSING
+            }
+
+        private fun constructNullValue(ionType: IonType = IonType.NULL) =
+            object : BaseExprValue() {
+                override val type = ExprValueType.NULL
+
+                @Suppress("UNCHECKED_CAST")
+                override fun <T> provideFacet(type: Class<T>?): T? = when (type) {
+                    IonType::class.java -> ionType as T?
+                    else -> null
+                }
+            }
+
+        private abstract class ScalarExprValue : BaseExprValue(), Scalar {
+            override val scalar: Scalar
+                get() = this
+        }
+
+        private fun constructBoolValue(value: Boolean) =
+            object : ScalarExprValue() {
+                override val type = ExprValueType.BOOL
+                override fun booleanValue(): Boolean = value
+            }
+
+        private fun constructStringValue(value: String) =
+            object : ScalarExprValue() {
+                override val type = ExprValueType.STRING
+                override fun stringValue() = value
+            }
```

Comment:
The original way makes `ExprValue` seem to have class inheritance structure, while actually it is not. It is just a union of different kotlin classes where each of them represents certain types of PartiQL scalar values.

Reply:
Why the `object` construction way makes the compiler harder to arrive?

Reply:
> Why the object construction way makes the compiler harder to arrive?

Because (1) singleton `object` is not a native concept in JVM bytecode / Java, which only has classes and (2) here, since these `object`s are created inside of constructXxx functions, all the `object`s that can be produced are not statically known.  So, your code is probably dealt with by inventing a class `Xxx` for each `constructXxx` method and implementing the `constructXxx`-produced `object` as an inner class of `Xxx`.  Then, maybe, the optimizer realizes that this nested structure is excessive and collapses `Xxx` and its nested inner class into one.  
So, at best, Kotlin (and/or JIT) works harder to get to the same result; at worst, it doesn't and the result is worse because of the nesting. 

In general, singleton `object`s are only useful on the top level, when they are true "static" singletons.  `object`-producing functions are probably ok in moderation, but here we are dealing with `ExprValue` which is *the* runtime of our system.

Reply:
I think here `object` is not used to specify a singleton, it is used to specify anonymous class

Reply:
Fine.  Substitute "singleton object" with "anonymous class" while reading the explanation while keeping in mind that there are no anonymous classes in the bytecode -- each class must have a statically-known name.

Reply:
I will revert the change